kadi/commands/observations.py

```diff
@@ -181,17 +181,82 @@ def convert_aostrcat_to_acatable(obs, params):
     return aca
 
 
-def get_starcats(obsid=None, *, start=None, stop=None, set_ids=True, scenario=None, cmds=None):
-    """Get star catalogs corresponding to input parameters.
+def get_starcats_as_table(start=None, stop=None, *, obsid=None, unique=False,
+                          scenario=None, cmds=None):
+    """Get a single table of star catalog entries corresponding to input parameters.
 
-    The ``obsid``, ``start``, and ``stop`` parameters serve as matching filters
+    This function calls ``get_starcats`` with the same parameters and then
+    concatenates the results into a single table for convenience. In addition
+    to the usual star catalog columns, the ``obsid`` and ``starcat_date`` are
+    included.
+
+    The ``unique`` parameter can be set to ``True`` to only return unique star
+    catalog entries. There are numerous instances of a single commanded star
+    catalogs that is associated with two ObsIDs, for instance ACIS undercover
+    observations. To get only the first one, set ``unique=True``.
+
+    In the following example we get every unique commanded guide star in 2020
+    and then join that with the corresponding observation information::
+
+      >>> from kadi.commands import get_starcats_as_table, get_observations
+      >>> from astropy import table
+      >>> start='2020:001'
+      >>> stop='2021:001'
+      >>> aces = get_starcats_as_table(start, stop, unique=True)
+      >>> ok = np.isin(aces['type'], ['GUI', 'BOT'])
+      >>> guides = aces[ok]
+      >>> obss = table.Table(get_observations(start, stop))
+      >>> obss = obss[~obss['starcat_date'].mask]  # keep only obs with starcat
+      >>> guides = table.join(guides, obss, keys=['starcat_date', 'obsid'])
+
+    :param start: CxoTime-like, None Start time (default=beginning of commands)
+    :param stop: CxoTime-like, None Stop time (default=end of commands)
+    :param obsid: int, None ObsID
+    :param unique: bool, if True return remove duplicate entries
+    :param scenario: str, None Scenario
+    :param cmds: CommandTable, None Use this command table instead of querying
+        the archive.
+    :returns: astropy ``Table`` star catalog entries for matching observations.
+
+    """
+    starcats = get_starcats(obsid=obsid, start=start, stop=stop, scenario=scenario,
+                            cmds=cmds, as_dict=True)
+    out = defaultdict(list)
+    for starcat in starcats:
+        n_cat = len(starcat['slot'])
+        out['obsid'].append([starcat['meta']['obsid']] * n_cat)
```

Comment:
Looks like this doesn't quite work if there isn't a cache already.

```
In [1]: import kadi.commands.observations

In [2]: dat = kadi.commands.observations.get_starcats_as_table(start='2022:001')
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-2-7adf682296db> in <module>
----> 1 dat = kadi.commands.observations.get_starcats_as_table(start='2022:001')

~/git/kadi/kadi/commands/observations.py in get_starcats_as_table(start, stop, obsid, unique, scenario, cmds)
    225     for starcat in starcats:
    226         n_cat = len(starcat['slot'])
--> 227         out['obsid'].append([starcat['meta']['obsid']] * n_cat)
    228         out['starcat_date'].append([starcat['meta']['date']] * n_cat)
    229         for name, vals in starcat.items():

~/miniconda3/envs/ska3/lib/python3.8/site-packages/astropy/table/table.py in __getitem__(self, item)
   1880     def __getitem__(self, item):
   1881         if isinstance(item, str):
-> 1882             return self.columns[item]
   1883         elif isinstance(item, (int, np.integer)):
   1884             return self.Row(self, item)

~/miniconda3/envs/ska3/lib/python3.8/site-packages/astropy/table/table.py in __getitem__(self, item)
    245         """
    246         if isinstance(item, str):
--> 247             return OrderedDict.__getitem__(self, item)
    248         elif isinstance(item, (int, np.integer)):
    249             return list(self.values())[item]

KeyError: 'meta'
```

Reply:
Yikes, not sure how this was passing tests since for testing the `~/.kadi` commands dir is set to a temp dir which would not have a `starcats.db` cache file. Anyway thanks for the catch of this pretty bad error, that's why we need independent testing! 

Here is my testing now:
```
(ska3) ➜  kadi git:(cache-starcat) ✗ rm ~/.kadi/starcats.db
(ska3) ➜  kadi git:(cache-starcat) ✗ ipython               
Python 3.8.12 (default, Oct 12 2021, 06:23:56) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.29.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from kadi.commands import *

In [2]: scs = get_starcats('2022:001')

In [4]: %time scs = get_starcats('2022:001', as_dict=True)
CPU times: user 66.6 ms, sys: 2.47 ms, total: 69 ms
Wall time: 68.2 ms

In [5]: %time scs = get_starcats('2022:001', as_dict=False)
CPU times: user 720 ms, sys: 5.36 ms, total: 725 ms
Wall time: 724 ms

In [6]: rm ~/.kadi/starcats.db

In [7]: %time scs = get_starcats('2022:001', as_dict=True)
CPU times: user 10.7 s, sys: 400 ms, total: 11.1 s
Wall time: 5.16 s

In [8]: %time scs = get_starcats('2022:001', as_dict=True)
CPU times: user 60.4 ms, sys: 2.57 ms, total: 63 ms
Wall time: 62.2 ms

In [9]: from kadi.commands import conf

In [10]: conf.cache_starcats = False

In [11]: rm ~/.kadi/starcats.db

In [12]: %time scs = get_starcats('2022:001', as_dict=True)
CPU times: user 10.6 s, sys: 384 ms, total: 11 s
Wall time: 5.08 s

In [13]: %time scs = get_starcats('2022:001', as_dict=True)
CPU times: user 10.6 s, sys: 367 ms, total: 11 s
Wall time: 5.13 s

In [14]: %time scs = get_starcats('2022:001', as_dict=False)
CPU times: user 11 s, sys: 368 ms, total: 11.3 s
Wall time: 5.45 s

In [15]: %time scs = get_starcats_as_table('2022:001')
CPU times: user 10.7 s, sys: 358 ms, total: 11.1 s
Wall time: 5.22 s

In [16]: ls ~/.kadi/starcats.db
ls: /Users/aldcroft/.kadi/starcats.db: No such file or directory
```